### PR TITLE
Feature/strong typed prop transmission target

### PR DESCRIPTION
**Pull request by jaelim-ms**
Typed property implementation for transmission targets.

### CODE REVIEW

File: AppCenterAnalytics/AppCenterAnalyticsTests/MSAnalyticsTransmissionTargetTests.m [hunk 1]
```diff
@@ -135,25 +176,28 @@ - (void)testTransmissionTargetForToken {
...
+  }
   NSString *event1 = @"event1";
   NSString *event2 = @"event2";
   NSString *event3 = @"event3";
   NSString *event4 = @"event4";
-
   MSAnalyticsTransmissionTarget *transmissionTarget, *transmissionTarget2;
-
+  OCMStub([self.analyticsClassMock canBeUsed]).andReturn(YES);
   // Events tracked when disabled mustn't be sent.
```
Comment (jwallra) [inline AppCenterAnalytics/AppCenterAnalyticsTests/MSAnalyticsTransmissionTargetTests.m]: New line before comment

File: AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m [hunk 2]
```diff
@@ -71,24 +72,36 @@ - (void)trackEvent:(NSString *)eventName {
...
     target = target.parentTarget;
   }
 
   // Override properties.
-  if (properties) {
-    [mergedProperties addEntriesFromDictionary:(NSDictionary * _Nonnull)properties];
-  } else if ([mergedProperties count] == 0) {
+  if (properties && properties.properties) {
+    [mergedProperties.properties addEntriesFromDictionary:(NSDictionary<NSString *, MSTypedProperty *> *)properties.properties];
+  } else if ([mergedProperties.properties count] == 0) {
```
Comment (clpolet) [inline AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m]: You can use isEmpty from MSEventProperties directly :)

File: AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m [hunk 3]
```diff
@@ -71,24 +72,36 @@ - (void)trackEvent:(NSString *)eventName {
...
+    [target.propertyConfigurator mergeTypedPropertiesWith:mergedProperties];
     target = target.parentTarget;
   }
 
   // Override properties.
-  if (properties) {
-    [mergedProperties addEntriesFromDictionary:(NSDictionary * _Nonnull)properties];
-  } else if ([mergedProperties count] == 0) {
+  if (properties && properties.properties) {
+    [mergedProperties.properties addEntriesFromDictionary:(NSDictionary<NSString *, MSTypedProperty *> *)properties.properties];
```
Comment (clpolet) [inline AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m]: Not a blocker for merging I would just add a `addEntriesFromEventProperties` to `MSEventProperties` so we don't use mergedProperties.properties directly.

File: AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSPropertyConfigurator.m [hunk 4]
```diff
@@ -60,7 +63,7 @@ - (void)removeEventPropertyForKey:(NSString *)propertyKey {
       MSLogError([MSAnalytics logTag], @"Event property key to remove cannot be nil.");
       return;
     }
-    [self.eventProperties removeObjectForKey:propertyKey];
+    [self.eventProperties.properties removeObjectForKey:propertyKey];
```
Comment (clpolet) [inline AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSPropertyConfigurator.m]: You can use the remove method form the MSEventProperties directly.

Comment (guperrot) [reply AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSPropertyConfigurator.m]: There is no remove method on MSEventProperties unless I am missing something @clpolet

Comment (clpolet) [reply AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSPropertyConfigurator.m]: Right I got confused.

File: AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m (hunk 2 above)
Comment (jaelim-ms) [reply AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m]: 👍

File: AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m (hunk 3 above)
Comment (jaelim-ms) [reply AppCenterAnalytics/AppCenterAnalytics/TransmissionTarget/MSAnalyticsTransmissionTarget.m]: OK

File: AppCenterAnalytics/AppCenterAnalyticsTests/MSAnalyticsTransmissionTargetTests.m (hunk 1 above)
Comment (jaelim-ms) [reply AppCenterAnalytics/AppCenterAnalyticsTests/MSAnalyticsTransmissionTargetTests.m]: Sure.

File: AppCenterAnalytics/AppCenterAnalytics/MSEventProperties.m [hunk 5]
```diff
@@ -62,7 +62,7 @@ - (instancetype)setDouble:(double)value forKey:(NSString *)key {
   if ([MSEventProperties validateKey:key]) {
 
     // NaN returns false for all statements, so the only way to check if value is NaN is by value != value.
-    if (value == (double)INFINITY || value != value) {
+    if (value == (double)INFINITY || value == -(double)INFINITY || value != value) {
```
Comment (ElektrojungeMS) [inline AppCenterAnalytics/AppCenterAnalytics/MSEventProperties.m]: wohoo, good catch!

**Pull request merged**